src/ast/passes/fold_literals.cpp

```diff
@@ -0,0 +1,450 @@
+    case Operator::LOR:
+      return static_cast<uint64_t>(left || right);
+    case Operator::INVALID:
+    case Operator::ASSIGN:
+    case Operator::INCREMENT:
+    case Operator::DECREMENT:
+    case Operator::LNOT:
+    case Operator::BNOT:
+      break;
+  }
+  LOG(BUG) << "Unexpected binary operator: " << static_cast<int>(op);
+  __builtin_unreachable();
+}
+
+Expression *LiteralFolder::visit(Cast &cast)
+{
+  visitAndReplace(&cast.expr);
+  return nullptr;
+}
+
+Expression *LiteralFolder::visit(Unop &op)
+{
+  if (auto *integer = dynamic_cast<Integer *>(op.expr)) {
+    if (op.op == Operator::BNOT) {
+      // Still positive.
+      return ast_.make_node<Integer>(~integer->value, Location(op.loc));
+    } else if (op.op == Operator::LNOT) {
+      // Still positive.
+      return ast_.make_node<Integer>(!integer->value, Location(op.loc));
+    } else if (op.op == Operator::MINUS) {
+      // Ensure that it is representable as a negative value.
+      if (integer->value >
+          static_cast<uint64_t>(std::numeric_limits<int64_t>::max()) + 1) {
+        op.addError() << "negative value will underflow";
+        return nullptr;
+      }
+      // Carefully make the conversion. We need to ensure that this does not
+      // overflow/underflow while doing the math.
+      if (integer->value <= 2) {
+        return ast_.make_node<NegativeInteger>(
+            -static_cast<int64_t>(integer->value), Location(op.loc));
+      } else {
+        int64_t value = -2;
+        value -= static_cast<int64_t>(integer->value - 2);
+        return ast_.make_node<NegativeInteger>(value, Location(op.loc));
+      }
+    }
+  } else if (auto *integer = dynamic_cast<NegativeInteger *>(op.expr)) {
+    if (op.op == Operator::BNOT) {
+      // Always positive.
+      return ast_.make_node<Integer>(static_cast<uint64_t>(~integer->value),
+                                     Location(op.loc));
+    } else if (op.op == Operator::LNOT) {
+      // Always positive.
+      return ast_.make_node<Integer>(static_cast<uint64_t>(!integer->value),
+                                     Location(op.loc));
+    } else if (op.op == Operator::MINUS) {
+      // Ensure that it doesn't overflow. We do this by ensuring that is
+      // representable as a positive number, casting, and then adding 1 to
+      // the unsigned form.
+      int64_t value = integer->value + 1;
+      value = -value;
+      return ast_.make_node<Integer>(static_cast<uint64_t>(value) + 1,
+                                     Location(op.loc));
+    }
+  }
+  return nullptr;
+}
+
+Expression *LiteralFolder::visit(Binop &op)
+{
+  visitAndReplace(&op.left);
+  visitAndReplace(&op.right);
+
+  // Check for string cases.
+  auto *str = dynamic_cast<String *>(op.left);
+  auto *other = op.right;
+  if (str == nullptr) {
+    str = dynamic_cast<String *>(op.right);
+    other = op.left;
+  }
+  if (str) {
+    // For whatever reason you are allowed to fold "foo"+3.
```

Comment:
Wait what? You mean you can already do this in bpftrace?

Reply:
Yep.

Reply:
Umm, can you give me an example?

Reply:
https://github.com/bpftrace/bpftrace/blob/master/src/ast/passes/semantic_analyser.cpp#L1327

Reply:
I think it's something to do with char * offset

Reply:
Ah, my bad - it was for positional params only 🙄